static/third/marked/lib/marked.js

```diff
@@ -1009,6 +1022,23 @@ Renderer.prototype.html = function(html) {
   return html;
 };
 
+Renderer.prototype.heading = function(text, level, raw, slugger) {
+  if (this.options.headerIds) {
+    return '<h'
+      + level
+      + ' id="'
+      + this.options.headerPrefix
+      + slugger.slug(raw)
+      + '">'
+      + text
+      + '</h'
+      + level
+      + '>\n';
+  }
```

Comment:
This feels like a really awkward way to do this; there must be a more idiomatic style we can do.

But in any case, I think we just need to remove `disable_markdown_regex(marked.Lexer.rules.tables, 'heading');` from `markdown.js` to enable the built-in feature.

Reply:
This was how upstream did it. Also, simply removing that line wouldn't help, since we had removed all the other code from marked that was supposed to add headings somewhere in the whole process of writing the frontend markdown.

The output below suggests that our `disable_markdown_regex` code was essentially no-op in this case:

```
$ git grep heading -- static/**/*.js
static/js/compose_ui.js:    // on what heading values have already been filled out.
static/js/confirm_dialog.js:        'html_heading',
static/js/confirm_dialog.js:    confirm_dialog.find('.confirm_dialog_heading').html(conf.html_heading);
static/js/emoji_picker.js:    popover.find('.emoji-popover-subheading').each(function () {
static/js/markdown.js:    // Disable headings
static/js/markdown.js:    disable_markdown_regex(marked.Lexer.rules.tables, 'heading');
static/js/markdown.js:    disable_markdown_regex(marked.Lexer.rules.tables, 'lheading');
static/js/portico/help.js:    // Highlight current article link and the heading of the same
static/js/portico/help.js:            // Close other article's headings first
static/js/portico/help.js:            // Toggle the heading
static/js/settings_user_groups.js:            html_heading: i18n.t('Delete user group'),
static/js/zform.js:                heading: schema.check_string,

```

Reply:
Actually, if we do not plan to reconcile our changes with upstream marked ever, we could ignore the whole `headerIds` codepath anyway, and turn this into a single line.

Reply:
5e5ca48063b4de3a6fb0d75202795756071b9412 is the commit that removed the heading code from marked.  I would prefer to restore that code (fixing the bug discussed in that commit) rather than adding a duplicate reimplementation; there's no reason for increase our divergence from upstream when adding a feature we want to work how upstream's works.

(We might need to tweak the regex or ordering or something to avoid conflicting with stream mentions, but that seems minor).

Reply:
I would like to clarify one thing: are we aiming to stay consistent with upstream as of that previous commit or a newer upstream?

I chose the heading regex and related code from upstream after some recent bug fixes related to this code were merged, and most of the part of the code is the same.

```
Old regex: /^ *(#{1,6}) *([^\n]+?) *#* *(?:\n+|$)/
New regex: /^ {0,3}(#{1,6}) +([^\n]*?)(?: +#+)? *(?:\n+|$)/
```

Main differences: Only allow up to 3 spaces at the start, require at least 1 space after #.

Reply:
That sounds great, so if that's what new upstream does, we should stick with it.  

(More generally, we should probably do a rebase of this with latest upstream).